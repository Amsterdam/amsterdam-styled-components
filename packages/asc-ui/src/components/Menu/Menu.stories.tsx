import React from 'react'
import { storiesOf } from '@storybook/react'
import { ReactComponent as Ellipsis } from '@datapunt/asc-assets/lib/Icons/Ellipsis.svg'
import { ReactComponent as FacebookPadded } from '@datapunt/asc-assets/lib/Icons/FacebookPadded.svg'
import { ReactComponent as Twitter } from '@datapunt/asc-assets/lib/Icons/Twitter.svg'
import { ReactComponent as Linkedin } from '@datapunt/asc-assets/lib/Icons/Linkedin.svg'
import { ReactComponent as Email } from '@datapunt/asc-assets/lib/Icons/Email.svg'
import { ReactComponent as Print } from '@datapunt/asc-assets/lib/Icons/Print.svg'
import { Menu, MenuItem, Icon } from '../..'

storiesOf('Menu', module)
  .add('default', () => (
    <Menu>
      <MenuItem role="button" onClick={() => {}}>
        One
      </MenuItem>
      <MenuItem role="button" onClick={() => {}}>
        Two
      </MenuItem>
    </Menu>
  ))
  .add('orientation top', () => (
    <Menu orientation="top">
      <MenuItem role="button" onClick={() => {}}>
        One
      </MenuItem>
      <MenuItem role="button" onClick={() => {}}>
        Two
      </MenuItem>
    </Menu>
  ))
  .add('with label', () => (
    <Menu label="I'm a label!">
      <MenuItem role="button" onClick={() => {}}>
        One
      </MenuItem>
      <MenuItem role="button" onClick={() => {}}>
        Two
      </MenuItem>
    </Menu>
  ))
  .add('example of actual usage', () => (
    <Menu
      orientation="top"
      icon={
        <Icon padding={4} inline size={24}>
          <Ellipsis />
        </Icon>
      }
    >
      <MenuItem
        role="button"
        divider
        onClick={() => {
          window.print()
        }}
        icon={
          <Icon padding={4} inline size={24}>
            <Print />
          </Icon>
        }
      >
        Printen
      </MenuItem>
      <MenuItem
        role="button"
        onClick={() => {
          window.open(
            `https://www.facebook.com/sharer/sharer.php?u=` +
              `${window.location.href}&title=${window.document.title}`,
            '_blank',
          )
        }}
        icon={
          <Icon inline size={24}>
            <FacebookPadded />
          </Icon>
        }
      >
        Deel via Facebook
      </MenuItem>
      <MenuItem
        role="button"
        onClick={() =>
          window.open(
            `https://twitter.com/intent/tweet?url=${window.location.href}` +
              `&text=${window.document.title}`,
            '_blank',
          )
        }
        icon={
          <Icon inline size={24} padding={4}>
            <Twitter />
          </Icon>
        }
      >
        Deel via Twitter
      </MenuItem>
      <MenuItem
        role="button"
        onClick={() =>
          window.open(
            `
          https://www.linkedin.com/shareArticle?url=${window.location.href}` +
              `&mini=true&title=${window.document.title}`,
            '_blank',
          )
        }
        icon={
          <Icon inline size={24} padding={4}>
            <Linkedin />
          </Icon>
        }
      >
        Deel via Linkedin
      </MenuItem>
      <MenuItem
        role="button"
        onClick={() => window.open(`mailto:mail@mail.com`, '_blank')}
        icon={
          <Icon inline size={24} padding={4}>
            <Email />
          </Icon>
        }
      >
        Deel via E-mail
      </MenuItem>
    </Menu>
  ))
