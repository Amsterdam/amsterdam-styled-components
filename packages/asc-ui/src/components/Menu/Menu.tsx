import React from 'react'
import { AscCore } from '@datapunt/asc-core'
import MenuButton from './MenuButton'
import MenuList from './MenuList'
import { KeyboardKeys } from '../../types'
import ownerDocument from '../../utils/ownerDocument'

type Props = {
  orientation?: 'top'
  label?: string
  icon?: React.ReactNode
}

type State = {}

const selectedChildInitial = -1

class Menu extends React.Component<Props, State> {
  state = {
    open: false,
    selectedChild: selectedChildInitial,
  }

  list = React.createRef<HTMLDivElement>()

  root = React.createRef<HTMLDivElement>()

  onKeyDown = (event: React.KeyboardEvent) => {
    const { children } = this.props
    const { selectedChild, open } = this.state

    const nrOfChildren = React.Children.count(children)

    if (!open) {
      return
    }

    const firstChild = 0
    const lastChild = nrOfChildren - 1

    if (event.key === KeyboardKeys.ArrowDown) {
      event.preventDefault()
      this.setState({
        selectedChild:
          selectedChild === lastChild ? firstChild : selectedChild + 1,
      })
    }

    if (event.key === KeyboardKeys.ArrowUp) {
      event.preventDefault()
      this.setState({
        selectedChild:
          selectedChild === firstChild || selectedChild === selectedChildInitial
            ? lastChild
            : selectedChild - 1,
      })
    }
  }

  onToggle = () => {
    const { open } = this.state
    this.setState({
      open: !open,
    })
  }

  onClose = () => {
    setTimeout(() => {
      const element = this.getReference('root') as HTMLInputElement
      if (element) {
        const currentFocus = ownerDocument(element).activeElement
        if (!element.contains(currentFocus)) {
          this.setState({
            selectedChild: selectedChildInitial,
            open: false,
          })
        }
      }
    })
  }

  getReference = (el: string) => {
    if (this[el].current) {
      return this[el].current
    }

    return null
  }

  render() {
    const { id, label, children, orientation, icon }: any = this.props
    const { open, selectedChild } = this.state

    return (
      <AscCore.Menu.MenuWrapper
        id={id}
        ref={this.root}
        onKeyDown={this.onKeyDown}
        onBlur={this.onClose}
      >
        <MenuButton
          {...{
            icon,
            open,
            orientation,
            label,
          }}
          onClick={this.onToggle}
        />
        <MenuList
          {...{
            orientation,
            id,
            open,
            selectedChild,
          }}
          onClose={this.onClose}
          ref={this.list}
        >
          {children}
        </MenuList>
      </AscCore.Menu.MenuWrapper>
    )
  }
}

export default Menu
