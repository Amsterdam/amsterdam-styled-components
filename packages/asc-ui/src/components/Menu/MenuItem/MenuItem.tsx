import React from 'react'
import MenuItemStyle from './MenuItemStyle'
import { Props as MenuItemLinkProps } from '../MenuItemLink/MenuItemLink'
import MenuItemLink from '../MenuItemLink'

type Props = {} & MenuItemLinkProps & React.HTMLProps<HTMLElement>

export default React.forwardRef<HTMLLIElement, Props>(function MenuItem(
  {
    children,
    linkIndex,
    element,
    onFocus,
    setCurrentLinkRef,
    iconRight,
    isActive,
    iconLeft,
    onClick,
    className,
    href,
    ...otherProps
  },
  ref,
) {
  return (
    <MenuItemStyle
      className={className}
      // @ts-ignore
      ref={ref}
      tabIndex={-1}
      {...otherProps}
    >
      <MenuItemLink
        {...{
          element,
          linkIndex,
          onFocus,
          onClick,
          setCurrentLinkRef,
          children,
          iconRight,
          isActive,
          iconLeft,
          href,
        }}
      >
        {children}
      </MenuItemLink>
    </MenuItemStyle>
  )
})
