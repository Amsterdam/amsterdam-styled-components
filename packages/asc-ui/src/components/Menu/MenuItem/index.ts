export { default as MenuItemStyle } from './MenuItemStyle'

export { default } from './MenuItem'
