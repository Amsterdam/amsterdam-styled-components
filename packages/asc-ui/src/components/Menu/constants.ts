export const FLY_OUT_WIDTH = 278
export const MENU_ITEM_SIZE = 50
