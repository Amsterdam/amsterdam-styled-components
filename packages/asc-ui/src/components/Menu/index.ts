import MenuDropDownStyle, { MenuDropDownButtonStyle } from './MenuDropDownStyle'
import MenuListStyle, { MenuListWrapperStyle } from './MenuListStyle'
import SubMenuListStyle, {
  SubMenuWrapperStyle,
  SubMenuListWrapperStyle,
} from './SubMenuListStyle'
import SubMenuButtonStyle, {
  SubMenuButtonLabelStyle,
} from './SubMenuButtonStyle'
import MenuButtonStyle from './MenuButtonStyle'
import MenuItemStyle, {
  MenuItemWrapperStyle,
  MenuItemLabelStyle,
} from './MenuItemStyle'
import MenuLabelStyle from './MenuLabelStyle'
import MenuBarStyle from './MenuBarStyle'

export { MenuStyleProps } from './types'

export const MenuStyle = {
  MenuDropDownStyle,
  MenuDropDownButtonStyle,
  MenuBarStyle,
  MenuListWrapperStyle,
  MenuListStyle,
  MenuItemStyle,
  MenuButtonStyle,
  MenuItemLabelStyle,
  MenuItemWrapperStyle,
  MenuLabelStyle,
  SubMenuWrapperStyle,
  SubMenuListWrapperStyle,
  SubMenuListStyle,
  SubMenuButtonStyle,
  SubMenuButtonLabelStyle,
}

export default MenuStyle
