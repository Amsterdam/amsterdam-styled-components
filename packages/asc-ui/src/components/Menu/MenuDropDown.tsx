import React from 'react'
import Icons from '@datapunt/asc-assets'
import MenuStyle, { MenuStyleProps } from './index'
import MenuList from './MenuList'
import { MenuContext } from './Menu'
import { Icon } from '../..'
import ReactIcon from '../ReactIcon/Icon'
import { KeyboardKeys } from '../../types'

const { MenuDropDownStyle, MenuDropDownButtonStyle } = MenuStyle

type Props = {
  position?: MenuStyleProps.Position
  label?: string
  mobile?: boolean
  arrowIcon?: React.ReactNode
  buttonHeight?: number
  id?: any
  index?: number
}

const MenuDropDown: React.FC<Props> = ({
  id,
  children,
  buttonHeight,
  position,
  index: currentIndex,
  mobile,
}) => {
  const menuRef = React.useRef<HTMLDivElement>(null)

  const { open, onKeyDown, onClose, onClick }: any = React.useContext(
    MenuContext,
  )

  const handleKeyPress = (e: React.KeyboardEvent) => {
    if (e.key === KeyboardKeys.Enter) {
      if (!open) {
        onClick()
      }
    } else {
      onKeyDown(e)
    }
  }

  const clonedChildren = React.Children.map(children, (child, index) =>
    React.cloneElement(child as React.ReactElement<any>, {
      index: currentIndex && currentIndex + index + 1,
    }),
  )

  const click = () => !mobile && setTimeout(onClick, 200)
  const clickWhenClosed = () => !mobile && !open && setTimeout(onClick, 200)

  return (
    <MenuDropDownStyle
      id={id}
      ref={menuRef}
      onKeyDown={handleKeyPress}
      onBlur={() => onClose(menuRef)}
      onMouseLeave={click}
      onMouseOver={clickWhenClosed}
      onFocus={clickWhenClosed}
      tabIndex={0}
    >
      <MenuDropDownButtonStyle
        {...{
          focused: open,
          height: buttonHeight,
        }}
        onMouseLeave={clickWhenClosed}
        onClick={onClick}
      >
        <Icon>
          <ReactIcon type={open ? Icons.Close : Icons.Menu} />
        </Icon>
      </MenuDropDownButtonStyle>
      <MenuList
        {...{
          position,
          id,
          open,
          top: buttonHeight,
        }}
        onClose={onClose}
      >
        {clonedChildren}
      </MenuList>
    </MenuDropDownStyle>
  )
}

MenuDropDown.defaultProps = {
  buttonHeight: 50,
}

export default MenuDropDown
