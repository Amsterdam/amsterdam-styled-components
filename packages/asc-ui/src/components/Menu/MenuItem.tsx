import React from 'react'
import MenuStyle, { MenuStyleProps } from './index'
import { Icon } from '../..'
import { KeyboardKeys } from '../../types'
import { useMenuContext } from './Menu'

const { MenuItemStyle, MenuItemWrapperStyle, MenuItemLabelStyle } = MenuStyle

type Props = {
  onClick?: Function
  role?: string
  icon?: React.ReactNode
  children?: any
  index?: number
  hoverColor?: string
  href?: string
  title?: string
} & MenuStyleProps.MenuItemStyleProps

const MenuItem = ({
  children,
  icon,
  index,
  onClick,
  hoverColor,
  href,
  title,
  ...otherProps
}: Props) => {
  const menuItemRef = React.useRef<HTMLAnchorElement>(null)

  const {
    selectedChild,
    onClick: onClickAction,
    expandedChild,
  }: any = useMenuContext()

  const focused = index === selectedChild

  React.useEffect(() => {
    if (menuItemRef && menuItemRef.current && focused) {
      menuItemRef.current.focus()
    }
  }, [selectedChild])

  const handleOnClick = (e: React.KeyboardEvent | React.MouseEvent) => {
    if (onClick) {
      onClickAction()
      onClick(e)
    }
  }

  const handleKeyPress = (e: React.KeyboardEvent) => {
    if (e.key === KeyboardKeys.Enter) {
      handleOnClick(e)
    }
  }

  return (
    <MenuItemWrapperStyle onKeyDown={handleKeyPress} tabIndex={-1}>
      <MenuItemStyle
        role="link"
        ref={menuItemRef}
        tabIndex={0}
        onClick={handleOnClick}
        hoverColor={expandedChild ? 'secondary' : hoverColor}
        onKeyDown={handleKeyPress}
        focused={focused}
        {...{ href, title }}
        {...otherProps}
      >
        {icon && <Icon size={14}>{icon}</Icon>}
        {/* CSS trick: we add an extra element span here so that MenuItemLabelStyle's
        border-bottom will underline overflowing text, since it's parent has display: flex */}
        <span>
          <MenuItemLabelStyle>{children}</MenuItemLabelStyle>
        </span>
      </MenuItemStyle>
    </MenuItemWrapperStyle>
  )
}

MenuItem.defaultProps = {
  height: 50,
  hoverColor: 'none',
}

export default MenuItem
