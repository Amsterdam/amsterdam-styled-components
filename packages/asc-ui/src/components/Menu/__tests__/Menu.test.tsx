import * as React from 'react'
import { mount } from 'enzyme'
import MenuItem from '../MenuItem'
import Menu, { MenuContext } from '../Menu'
import MenuDropDown from '../MenuDropDown'

describe('Menu', () => {
  const mockOnClick = jest.fn()
  const mockOnKeyDown = jest.fn()

  const mockContext = {
    open: false,
    onKeyDown: mockOnKeyDown,
    onClick: mockOnClick,
  }

  const children = (
    <>
      <MenuItem />
      <MenuItem />
    </>
  )

  it('should return the correct children', () => {
    const component = mount(
      <MenuContext.Provider value={mockContext}>
        <Menu>{children}</Menu>
      </MenuContext.Provider>,
    )

    expect(component.props().children).toEqual(children)
  })

  it('should render the component for the mobile menu', () => {
    const component = mount(
      <MenuContext.Provider value={mockContext}>
        <Menu mobile>{children}</Menu>
      </MenuContext.Provider>,
    )

    expect(component.find(MenuDropDown).exists()).toBe(true)
  })
})
