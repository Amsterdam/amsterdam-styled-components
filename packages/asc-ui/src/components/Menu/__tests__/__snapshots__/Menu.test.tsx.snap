// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should render 1`] = `
<styled.div
  onBlur={[Function]}
  onKeyDown={[Function]}
>
  <MenuButton
    label="Click on me"
    onClick={[Function]}
    open={false}
  />
  <ForwardRef
    onClose={[Function]}
    open={false}
    selectedChild={-1}
  >
    <MenuItem>
      One
    </MenuItem>
    <MenuItem>
      Two
    </MenuItem>
    <MenuItem>
      Three
    </MenuItem>
  </ForwardRef>
</styled.div>
`;
